language/diem-framework/modules/Diem.move

```diff
@@ -485,97 +551,337 @@ module Diem {
+        modifies global<Preburn<CoinType>>(account_addr);
+        // Publishes if it doesn't exists. Updates its state either way.
+        modifies global<PreburnQueue<CoinType>>(account_addr);
+        // The preburn amount in the currency info can be updated.
+        modifies global<CurrencyInfo<CoinType>>(CoreAddresses::CURRENCY_INFO_ADDRESS());
         include PreburnToAbortsIf<CoinType>{amount: coin.value};
-        include PreburnEnsures<CoinType>{preburn: preburn, amount: coin.value};
+        include PreburnEnsures<CoinType>{preburn: spec_make_preburn(coin.value), amount: coin.value};
     }
-
     spec schema PreburnToAbortsIf<CoinType> {
         account: signer;
         amount: u64;
         let account_addr = Signer::spec_address_of(account);
-        let preburn = global<Preburn<CoinType>>(account_addr);
-        /// Must abort if the account does have the Preburn resource [[H4]][PERMISSION].
+        /// Must abort if the account doesn't have the PreburnQueue or Preburn resource [[H4]][PERMISSION].
+        /// This aborts condition is covered in the `UpgradePreburnAbortsIf` schema.
         include Roles::AbortsIfNotDesignatedDealer;
-        include AbortsIfNoPreburn<CoinType>{preburn_address: account_addr};
-        include PreburnWithResourceAbortsIf<CoinType>{preburn: preburn};
+        include PreburnAbortsIf<CoinType>;
+        include UpgradePreburnAbortsIf<CoinType>;
+        include AddPreburnToQueueAbortsIf<CoinType>{preburn: spec_make_preburn(amount)};
+    }
+
+    /// Remove the oldest preburn request in the `PreburnQueue<CoinType>`
+    /// resource published under `preburn_address` whose value is equal to `amount`.
+    /// Calls to this function will fail if:
+    /// * `preburn_address` doesn't have a `PreburnQueue<CoinType>` resource published under it; or
+    /// * a preburn request with the correct value for `amount` cannot be found in the preburn queue for `preburn_address`;
+    fun remove_preburn_from_queue<CoinType>(preburn_address: address, amount: u64): Preburn<CoinType>
+    acquires PreburnQueue {
+        assert(exists<PreburnQueue<CoinType>>(preburn_address), Errors::not_published(EPREBURN_QUEUE));
+        // We search from the head of the queue
+        let index = 0;
+        let preburn_queue = &mut borrow_global_mut<PreburnQueue<CoinType>>(preburn_address).preburns;
+        let queue_length = Vector::length(preburn_queue);
+
+        while ({
+            spec {
+                assert index <= queue_length;
+                assert forall j in 0..index: preburn_queue[j].to_burn.value != amount;
+            };
+            (index < queue_length)
+            }) {
+            let elem = Vector::borrow(preburn_queue, index);
+            if (value(&elem.to_burn) == amount) {
+                let preburn = Vector::remove(preburn_queue, index);
+                // Make sure that the value is correct
+                assert(value(&preburn.to_burn) == amount, Errors::invalid_state(EPREBURN));
```

Comment:
could this ever happen?

Reply:
Good spot. I don't think so...I'll remove